src/main/java/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt

```
@@ -56,17 +57,26 @@ open class MfaService(@Value("\${application.authentication.mfa.whitelist}") whi
     return Pair(token, code)
   }
 
-  @Transactional(transactionManager = "authTransactionManager")
-  open fun validateAndRemoveMfaCode(token: String, code: String?): Optional<String> {
-    if (code.isNullOrBlank()) return Optional.of("missingcode")
+  @Transactional(transactionManager = "authTransactionManager", noRollbackFor = [LoginFlowException::class, MfaFlowException::class])
```

@mikehalmamoj I've switched from `Optional` to exception processing, see what you think.
Had to therefore exclude the exceptions from the rollback here makes it a bit messier

---

Why is the rollback cancelled for the MfaFlowException?  Looks like it's just the incrementRetriesAndLock... that needs to happen?  Anyway, I think runtime exceptions are better where "fail fast" makes sense and you can use them with a ControllerAdvice, but in this situation failure requires a side effect (locking the account) so it's going to be messy either way.

---

When the user gets the mfa code incorrect we don't need to lock the account, but have incremented the retries i.e. we drop down to the `throw it` in
```java
      if (it.error.equals("invalid")) {
        val locked = userRetriesService.incrementRetriesAndLockAccountIfNecessary(userPersonDetails)
        if (locked) throw LoginFlowException("locked")
      }
      throw it
```